### ISSUE TITLE
Addon "forgets" VPN settings occasionally

### ISSUE THREAD

**Issue by Rippert** [Zomboided/service.vpn.manager#170]
Twice now, I've had the addon lose the VPN settings. I have to go through the wizard again and set it up each time. This is on LibreElec on an RPi, so I think the logs are going to a temp folder that gets wiped when the system restarts. 

Is this a know issue? If not, how should I set the settings so that I can get you a reasonable log. It only happens occasionally, so I would like to avoid running the main KODI debug log all the time if I can.

**Comment by Zomboided** [comment]
I've seen this before on older versions, not on the newer versions. Uninstall what you have in case there's something hanging around from earlier, then install the latest from the repo. Run with Advanced/Debug switched on.

Logs are in temp as you've noticed. On reboot the old log is renamed to kodi.old.log. Get me this and kodi.log if you see this happen again.

If you're using the repo and this is happening when a version gets upgraded then you'll need to wait for that to happen - which will probably be in the next few days.

**Comment by kuitro** [comment]
I can confirm this seems to be a problem/bug. I'm experiencing this since I set this up on a new RPi 3b+ with LibreElec about a week ago. The wizard gets displayed and the settings file is virgin.

I've been trying various things like making sure the RPi has network before Kodi starts with and without starting OpenVPN before Kodi starts, reinstalling the add-on. Nothing makes any difference, VPN settings just get lost now and again.

I still have to gather logs when this happens...

Right now as an ugly workaround I copy the correct settings.xml from "/storage/" to "/storage/.kodi/userdata/addon_data/service.vpn.manager/" at boot through "/storage/.config/autostart.sh".

**Comment by Zomboided** [comment]
What level of code are you using?  Is this some non-official build? (in
which case, recreate on a proper released build)
Can you get settings.xml with the debug log too before you copy over it.

**Comment by kuitro** [comment]
I'm using the current official build (LibreELEC-RPi2.arm-8.2.5).
Just in case this wasn't clear/obvious, this only happens sometimes after a reboot/coldboot.
I'm a little strapped for time so it might take a bit until I can provide logs.

What about Rippert's logs, was he able to provide one? Also, it would be good to know what build of LibreElec and RPi he is using.

**Comment by kuitro** [comment]
Oh and vpn manager 4.9.9 installed through the repo. Kodi is 17.6.

**Comment by Zomboided** [comment]
I've not got any logs from anyone showing this behaviour yet.  You're using
exactly the config I am though, so it's properly weird that it's happening.

**Comment by Rippert** [comment]
Sorry I haven't responded in a while. I did have one more loss of data, but it was before I turned on the logging in the VPN manager. Also before uninstalling and reinstalling from the repo. The only thing I noticed in the log at the time of the loss was that the PVR was throwing errors. I have since turned off all PVR functions.

I'm using LibreELEC 8.2.5, Kodi 17.6 Git:a9a7a20, and VPN manager 4.9.9

I'm not sure what version I had prior to the reinstall, but it was from the repo. I've got the log on now and will get you a log if it happens again.

**Comment by Rippert** [comment]
I did get another loss of setup data happen. The funny thing is that the VPN connection was still active. It seems that what is being lost is the Provider name, username and password, not the cert files. I never have to relocate them for the addon. 

The only obvious issue that I could find from the log is this:

> 
> 00:09:20.674 T:1710814112   ERROR: EXCEPTION: XBMC is not playing any file
> 00:09:20.694 T:1710814112   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
>                                              - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
>                                             Error Type: <type 'exceptions.RuntimeError'>
>                                             Error Contents: XBMC is not playing any file
>                                             Traceback (most recent call last):
>                                               File "/storage/.kodi/addons/service.vpn.manager/service.py", line 167, in onPlayBackStarted
>                                                 filename = self.getPlayingFile()
>                                             RuntimeError: XBMC is not playing any file
>                                             -->End of Python script error report<--

Let me know your preferred method to upload the full log file. I'll be traveling for the next couple days, so I may not get it back to you right away.

**Comment by Zomboided** [comment]
Upload the new and the old logs to pastebin

**Comment by Rippert** [comment]
Here is the log when the VPNManager "forgot":
[https://pastebin.com/mRBUXzYk](url)

I used the "copy log file" option in VPNManager, which only copied the current log, and I've rebooted several time since then. Sorry, but I'm very sure that the VPNManager "knew" the account details and then "forget" them all in the one session covered in this log. If it happens again, I'll get both new and old logs.

**Comment by Zomboided** [comment]
This log shows it connecting at boot and reconnecting when your vpn provider decides to disconnect you. If the details had been wiped out it wouldn't have connected at boot. I can't do anything with this log as it doesn't show the behaviour you're claiming.

**Comment by Rippert** [comment]
That's what I'm trying to tell you. The VPN details were there when I booted, and continued to be there for some time, but at one point, when I went to check the connection, the addon brought up a dialog saying that there weren't any VPNs setup, and asking me to use the wizard to set them up. When I chose to go into "settings" instead of "wizard" the VPN name, username and password were all blank.

This all happened during the one session that this log covers.

If you have some suggestion for how to get something useful into the log, let me know.

**Comment by Zomboided** [comment]
I've had another look and I don't see where this is happening.  There's no trace of you trying to reconnect via the settings, but there's evidence of the VPN working at the start, and just before you gathered the log at the end and a couple of normal looking reconnects in between.

I've pushed out a new version (5.0.3) which fixes the exception in the log and adds some more trace when you're asked for the wizard.  You can get this from GitHub or the repo when it refreshes.  

You should install 5.0.3 and continue to run with debug on.  It would be helpful to know the time you think the problem happened - I don't want to trawl through 1000s of lines of trace trying to guess.  If you see it again, get the log immediately (I don't need the old one) and put it up on pastebin again.

**Comment by Zomboided** [comment]
I never pushed 5.0.1 to the repo, or 5.0.2

If you find the Zomboided repo in the Systems/Addons (depending on the skin
you're using) and force an update for it, it should find it.

**Comment by Rippert** [comment]
I just checked and it upgraded to 5.0.3 automatically. I had to revalidate the VPN, and reenable the debug log setting. I'll let you know if I have another loss of settings.

**Comment by Zomboided** [comment]
Ok thanks - the revalidation was because I pushed a CyberGhost update this
afternoon too, which I noticed you were using.  It was just for the
Netherlands, which you probably weren't using otherwise you'd have been
telling me that didn't work too :-)

**Comment by dannyshaw** [comment]
I can confirm this behaviour. I have openelec Kodi installed last night and this add-on has lost its settings and reprompted the wizard 2 times since initial setup over about 3 hours.

**Comment by Zomboided** [comment]
There's no point in confirming it without getting the debug log before and after it happens that I've been asking for...I'm not debating that this an issue for some people!

**Comment by Raylan189** [comment]
I can also confirm that this problem still exists on the Raspberry Pi. My environment is a bit different:
- Raspbian Stretch instead of LibreELEC,
- Kodi 17.6 installed afterwards with APT,
- addon version 5.0.3 with VPN User Defined.

I have no debug files for you, but luckily I have made a backup of the settings file just before the error occured. From the fact that my backup file was not empty I conclude that the error happens on program start instead of on program exit (as I previously expected). It seems to me that the addon fails to read the existing settings file and writes an empty file back later on. If this information is new for you, I hope hope that it helps you fix the problem.

**Comment by Zomboided** [comment]
Nope, I need to see logs...

**Comment by gianmarcotoso** [comment]
Can confirm, this just happen on my installation (LibreELEC with Kodi 18, VPNManager v 5.0.6) after the VPN randomly disconnected. The VPN itself kept connecting on startup, but VPNManager kept asking to be reconfigured only to fail during configuration. Settings screen was also inaccessible (blank window), so I solved by reinstalling and reconfiguring. Next time I'll gather the logs and post them here, didn't think about it this time, sorry :(

**Comment by Zomboided** [comment]
@gianmarcotoso I'm wondering if this is related to a disconnect happening - what was the chain of events?

You were connected and everything was good.
A disconnect happened because of some external network reason.
Did it then reconnect, or did it continue to fail to reconnect?
And did you see the problem only after you rebooted?

I understand you reinstalled then and everything was fine.

**Comment by gianmarcotoso** [comment]
As far as I can tell it did not reconnect after the first disconnection (I sshd into the box and checked my IP against a remote service and it was my public static IP, so I was definitely not connected), and I wasn't able to reconfigure it neither then nor after rebooting. I don't know if it did try to reconnect, but I have the "ask to reboot after 10 failures" setting on so it should have told me to reboot if it continued to fail. Unless the bug is related to that option?

The wizard failed when "writing credentials in the ovpn" (an error along those lines, I was kind of in a hurry and did not write it down, sorry...) and the settings screen was broken. But yes, the chain of events you described is how it happened.

**Comment by Zomboided** [comment]
@gianmarcotoso Are you using a RPi or some other box that allows you to connect to the VPN before Kodi boots?  I'm guessing so based on your comments but need to check.

**Comment by gianmarcotoso** [comment]
You are absolutely correct, RPi Zero with LibreElec 18 (Milhouse build #0920), and the option to connect before Kodi starts is enabled in OpenVPN Manager.

**Comment by gianmarcotoso** [comment]
A little update: it happened again, and this time I have more info. The VPN got disconnected because of inactivity, and for some reason the openvpn process received a SIGKILL signal and died.

I think I'm now able to reproduce it:

- Reinstall and reconfigure the Kodi addon
- Reboot the system so that openvpn autoconnects before boot
- Idle
- Wait for the "missing configuration" window to pop up.

**Comment by Zomboided** [comment]
This is some good news!  I looked at the code and couldn't see where the
settings were being blatted.  Knowing how the disconnect is detected would
be really helpful.

Go to settings, switch on debug in the advanced menu.

When you see the error, get me the kodi.log and the old.kodi.log.  They'll
be in /storage/.kodi/temp/
This gets me the log before the reboot and the log after.  I suspect the
current log will tell me what's happening.

Also in /storage/.kodi/userdata/addon_data/service.vpn.manager/ is a file
called settings.xml.
If you can get that *before* the idle happens and after you've seen the
error, that'll be helpful too.  Before you share these, remove any userID
and passwords - don't delete them, just xxxxxx them out if they're there,
I'll need to know if they're stored.  If there are none, just leave the
values as a blank field.

Finally, maybe take a picture of the error that pops up too, just in case I
don't trace it.

I can see from the logs, but who's your VPN provider?

**Comment by gianmarcotoso** [comment]
Sorry, I've been a little busy with work and haven't been able to gather the data you requested. Meanwhile the RPI has been used a little bit, so the logs are not "pristine". The current status is that the addon complains about not being a configuration, but the VPN is correctly working with `openvpn`. If I cancel the wizard (I press back until it stops complaining) everything works correctly (except for the addon). Would you like me to gather the logs in this state, or should I reinstall the plugin and force the issue from scratch?

BTW My VPN provider is Mullvad, but I also have NordVPN.

**Comment by Zomboided** [comment]
Make sure debug is on, go through the wizard and then get me the current log and the settings.xml

**Comment by grant2258** [comment]
Im having odd behavior on libreelec 8.2.5 when i boot the vpn says its connected and doesnt work. If disconnect from a vpn manually and reconnect through the plugin via change or disconnect VPN connection  it works properly again. I have to repeat this again If i leave the the box on too long and it needs to wake up.

**Comment by Zomboided** [comment]
Get a log with debug, open a separate issue - it's not the same as this
issue.

**Comment by grant2258** [comment]
Well the issue is fixed as of release 5.0.6 so my issue is fixed

**Comment by doncorso** [comment]
Hello, I am using the libreelecs milhouse build from this week on a raspberry 3
I have never used your vpn-manager before and try using it the first time.
The problem, I encountered, is: when I try establishing a vpn connection I get an error, and the wizard wants to help me again and again.
kodi log shows: 
```
ERROR: EXCEPTION Thrown (PythonToCppException): -->Pyton callack/script returned the following error<--
NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
Error type: <type 'exceptions.ValueError'>
Error Contents: " is not in list
Traceback (most recent call last):
File /storage/.kodi(addons/sservice.vpn.manager/libs/common.py", line 1153, in wizard
 i = provider_list.index(getVPNDisplay(current))
ValueError: "is not in list

```

Thanks for your help

DC

.

**Comment by Zomboided** [comment]
@doncorso This problem is differnt to the issue you've commented on.
Also, you've not provided a log.  You've not stated which level of code you're on.  And I'm not interested in debugging issues against nightly builds which are inherintly unstable.
So recreate on a released Kodi 18 Leia with the latest VPN Mgr build, and if there's still an issue, raise a new issue with a FULL debug log.

**Comment by doncorso** [comment]
Thank you for your fast answer. I managed it, using ovpn-files from my vpn-provider.
Works perfect.
I say Thank you for your work here.

Thumbs up!

**Comment by Raylan189** [comment]
I only want to inform you that the addon now runs perfectly stable on my Raspberry Pi with Raspbian Stretch. The settings file has not been overwritten for months now. I have not upgraded my system software until February 2019, so an addon change must be responsible for the bug fix.

From my point of view this request can be closed now.
Thank you very much for providing this addon, great work!

**Comment by Zomboided** [comment]
I've seen this on Kodi 18, and another user has as well.  The settings are completely erased/corrupted rather than the add-on deleting them.

**Comment by Zomboided** [comment]
Closing this as I've now wrapped some waits around the setting and re-accessing the settings.xml calls.  Any future reports should be against the latest Kodi with the latest add-on (so at least Kodi 18 and add-on 6.2.2)
